incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,148 @@
+        self.limit_of_revealed = limit_of_revealed
+
+        # map metacoalition size rank -> metacoalition id
+        self.meta_rank_to_id = metacoalition_ids_by_coalition_size(number_of_players, limit_of_revealed)
+        # these are the metacoalitions of size <= limit_of_revealed
+        self.viable_metacoalitions = len(self.meta_rank_to_id)
+        # map metacoalition id -> metacoalition size rank
+        self.meta_id_to_rank = np.zeros(self.viable_metacoalitions, dtype=int)
+        self.meta_id_to_rank[self.meta_rank_to_id] = np.arange(self.viable_metacoalitions)
+
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def get_actions_distribution(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
+        """Get the action predicted by the strategy.
+
+        If past_actions is an int, it is considered a metacoalition id.
+        Otherwise, it is considered as a list of coalitions of the original game.
+        """
+        metacoalition = past_actions if isinstance(past_actions, int) else self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        if sum(self.regret[rank]) == 0:
+            return np.ones(self.number_of_coalitions) / self.number_of_coalitions  # TODO: not all viable!
+        return self.regret[rank] * (self.regret[rank] > 0) / sum(self.regret[rank] * self.regret[rank] > 0)
+
+    def apply_regret(self, bottom_losses: np.ndarray, used_actions: list[list[Coalition]]) -> None:
+        """Scan every possible action sequence, and update the regret of all the regret minimizers.
+
+        Arguments:
+            bottom_losses: the loss of the bottom layer of the game tree.
+                    this should be an array of size (number_of_coalitions,).
+            used_actions: the list of list of actions used to generate the bottom_losses.
+        """
+        used_metacoalition_ranks = [self.meta_id_to_rank[self.get_metacoalition_id(x)]
+                                    for x in used_actions]
+        losses = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                          dtype=RMValue)
+        experienced_losses = np.zeros(self.viable_metacoalitions, dtype=RMValue)
```

Comment:
`meta_values` instead of `experienced_losses`?

Reply:
Sorry, but `meta_values` is not descriptive at all, I'm keeping this, because I wouldn't be able to read the code myself if I changed this :D

Reply:
Sure, I just wanted to push the meta in there somewhere to make it clear it is over meta-coalitions.